src/plugins/intel_npu/src/plugin/src/metadata.cpp

```diff
@@ -0,0 +1,123 @@
+    this->size = static_cast<uint32_t>(version.size());
+}
+
+void OpenvinoVersion::read(std::istream& stream) {
+    stream.read(reinterpret_cast<char*>(&size), sizeof(size));
+    stream.read(&version[0], size);
+}
+
+Metadata<METADATA_VERSION_1_0>::Metadata() : ovVersion{ov::get_openvino_version().buildNumber} {
+    version = METADATA_VERSION_1_0;
+}
+
+void Metadata<METADATA_VERSION_1_0>::read(std::istream& stream) {
+    ovVersion.read(stream);
+}
+
+uint32_t& OpenvinoVersion::get_size() {
+    return size;
+}
+
+void Metadata<METADATA_VERSION_1_0>::write(std::ostream& stream) {
+    stream.write(reinterpret_cast<const char*>(&version), sizeof(version));
+    stream.write(reinterpret_cast<const char*>(&ovVersion.get_size()), sizeof(ovVersion.get_size()));
+    stream.write(ovVersion.get_version().data(), ovVersion.get_version().size());
+}
+
+std::unique_ptr<MetadataBase> create_metadata(uint32_t version) {
+    switch (version) {
+    case METADATA_VERSION_1_0:
+        return std::make_unique<Metadata<METADATA_VERSION_1_0>>();
+
+    default:
+        return nullptr;
+    }
+}
+
+std::string_view OpenvinoVersion::get_version() {
+    return std::string_view(version);
+}
+
+bool Metadata<METADATA_VERSION_1_0>::is_compatible() {
+    Logger _logger("NPUPlugin", Logger::global().level());
+    // checking if we can import the blob
+    if (ovVersion.get_version().compare(ov::get_openvino_version().buildNumber)) {
+        _logger.warning("Imported blob OpenVINO version: %s, but the current OpenVINO version is: %s",
+                        ovVersion.get_version().data(),
+                        ov::get_openvino_version().buildNumber);
+
+#ifdef NPU_PLUGIN_DEVELOPER_BUILD
+        if (auto envVar = std::getenv("NPU_DISABLE_VERSION_CHECK")) {
+            if (envVarStrToBool("NPU_DISABLE_VERSION_CHECK", envVar)) {
+                return true;
+            }
+        }
+#endif
+        return false;
+    }
+    return true;
+}
+
+std::unique_ptr<MetadataBase> read_metadata_from(std::vector<uint8_t>& blob) {
+    Logger _logger("NPUPlugin", Logger::global().level());
+    size_t magicBytesSize = MAGIC_BYTES.size();
+    std::string blobMagicBytes(magicBytesSize, '\0');
+
+    auto metadataIterator = blob.end() - magicBytesSize;
+    memcpy(blobMagicBytes.data(), &(*metadataIterator), magicBytesSize);
+    if (MAGIC_BYTES != blobMagicBytes) {
+        _logger.error("Blob is not versioned!");
+        return nullptr;
+    }
+
+    size_t blobDataSize;
+    metadataIterator -= sizeof(blobDataSize);
+    memcpy(&blobDataSize, &(*metadataIterator), sizeof(blobDataSize));
+    metadataIterator = blob.begin() + blobDataSize;
+
+    std::stringstream metadataStream;
+    metadataStream.write(reinterpret_cast<const char*>(&(*metadataIterator)),
+                         blob.end() - metadataIterator - sizeof(blobDataSize));
+
+    uint32_t metaVersion;
+    metadataStream.read(reinterpret_cast<char*>(&metaVersion), sizeof(metaVersion));
```

Comment:
Should we check the major version is compatible before attempting to read the rest?

Also, why you saying in PR's description that removing the OV version would break backward compatibility with the old blobs? Did you actually mean forward compatibility? I.e. old plugin + new blob missing this field.

Reply:
The major version is already checked inside `create_metadata`. This implies that **any** metadata appended to the blob, has its version as the first field.

Reply:
Then shouldn't we reinforce forward compatibility in the code as well? As it stands, you're throwing if the plugin sees version 1.1 (a future metadata implementation), but you're description suggests this scenario should work.

Reply:
I'm not sure which part of the description implies this scenario should work.

Today's requirements say that we should throw unless metadata version and ov version match. (and display any helpful information through the logger)